syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "protos/assets.proto";

message Order {
    string id = 1;
    google.protobuf.Timestamp dt = 2;
    string reason = 3;
    google.protobuf.Timestamp created = 4;
    Asset asset = 5;
    float amount = 6;
    int32 filled = 7;
    float commission = 8;
    enum Status {
        OPEN = 0;
        FILLED = 1;
        CANCELED = 2;
        REJECTED = 3;
        HELD = 4;
    }
    Status status = 14;
    message Stop {
        bool is_set = 1;
        float value = 2;
    }
    message Limit {
        bool is_set = 1;
        float value = 2;
    }
    Stop stop = 9;
    Limit limit = 10;
    bool stop_reached = 11;
    bool limit_reached = 12;
    int64 broker_order_id = 13;
}

message Position {
    Asset asset = 1;
    int64 amount = 2;
    float cost_basis = 3;
    float last_sale_price = 4;
    google.protobuf.Timestamp last_sale_date = 5;
}

message Account {
    //todo: add fields for commodities...
    float settled_cash = 1;
    float accrued_interest = 2;
    float buying_power = 3;
    float equity_with_loan = 4;
    float total_positions_value = 5;
    float total_positions_exposure = 6;
    float regt_equity = 7;
    float regt_margin = 8;
    float initial_margin_requirement = 9;
    float maintenance_margin_requirement = 10;
    float available_funds = 11;
    float excess_liquidity = 12;
    float cushion = 13;
    float day_trades_remaining = 14;
    float leverage = 15;
    float net_leverage = 16;
    float net_liquidation = 17;
    float sma = 18;
}

message Portfolio {
    float cash_flow = 1;
    float starting_cash = 2;
    float portfolio_value = 3;
    float pnl = 4;
    float returns = 5;
    float cash = 6;
    repeated AssetPositionPair positions = 7;
    google.protobuf.Timestamp start_date = 8;
    float positions_value = 9;
    float positions_exposure = 10;
}

message AssetPositionPair {
    Asset key = 1;
    Position position = 2;
}